src/transactions/LiquidityPoolDepositOpFrame.cpp

```diff
@@ -0,0 +1,318 @@
+    int64_t& amountA, int64_t& amountB, int64_t& amountPoolShares,
+    int64_t availableA, int64_t availableB, int64_t availableLimitPoolShares,
+    LiquidityPoolConstantProduct const& cp)
+{
+    int64_t sharesA = 0;
+    int64_t sharesB = 0;
+    {
+        bool resA = bigDivide(sharesA, cp.totalPoolShares,
+                              mLiquidityPoolDeposit.maxAmountA, cp.reserveA,
+                              ROUND_DOWN);
+        bool resB = bigDivide(sharesB, cp.totalPoolShares,
+                              mLiquidityPoolDeposit.maxAmountB, cp.reserveB,
+                              ROUND_DOWN);
+        if (!minAmongValid(amountPoolShares, sharesA, resA, sharesB, resB))
+        {
+            // sharesA > INT64_MAX and sharesB > INT64_MAX so we always fail
+            innerResult().code(LIQUIDITY_POOL_DEPOSIT_POOL_FULL);
+            return false;
+        }
+    }
+
+    {
+        bool resA = bigDivide(amountA, amountPoolShares, cp.reserveA,
+                              cp.totalPoolShares, ROUND_UP);
+        bool resB = bigDivide(amountB, amountPoolShares, cp.reserveB,
+                              cp.totalPoolShares, ROUND_UP);
+        if (!(resA && resB))
+        {
+            innerResult().code(LIQUIDITY_POOL_DEPOSIT_UNDERFUNDED);
+            return false;
+        }
+    }
+
+    if (availableA < amountA || availableB < amountB)
+    {
+        innerResult().code(LIQUIDITY_POOL_DEPOSIT_UNDERFUNDED);
+        return false;
+    }
+
+    if (isBadPrice(amountA, amountB, mLiquidityPoolDeposit.minPrice,
+                   mLiquidityPoolDeposit.maxPrice))
+    {
+        innerResult().code(LIQUIDITY_POOL_DEPOSIT_BAD_PRICE);
+        return false;
+    }
+
+    if (availableLimitPoolShares < amountPoolShares)
+    {
+        innerResult().code(LIQUIDITY_POOL_DEPOSIT_LINE_FULL);
+        return false;
+    }
+
+    return true;
+}
+
+static void
+updateBalance(LedgerTxnHeader& header, TrustLineWrapper& tl,
+              LedgerTxnEntry& acc, int64_t delta)
+{
+    if (tl)
+    {
+        if (!tl.addBalance(header, delta))
+        {
+            throw std::runtime_error("insufficient balance");
+        }
+    }
+    else
+    {
+        if (!stellar::addBalance(header, acc, delta))
+        {
+            throw std::runtime_error("insufficient balance");
+        }
+    }
+}
+
+bool
+LiquidityPoolDepositOpFrame::doApply(AbstractLedgerTxn& ltx)
+{
+    // Don't need TrustLineWrapper here because pool share trust lines cannot be
+    // issuer trust lines.
+    auto tlPool = loadPoolShareTrustLine(ltx, getSourceID(),
+                                         mLiquidityPoolDeposit.liquidityPoolID);
+    if (!tlPool)
+    {
+        innerResult().code(LIQUIDITY_POOL_DEPOSIT_NO_TRUST);
+        return false;
+    }
+
+    // lp must exist if tlPool exists
+    auto lp = loadLiquidityPool(ltx, mLiquidityPoolDeposit.liquidityPoolID);
+    auto& cp = lp.current().data.liquidityPool().body.constantProduct();
```

Comment:
We started trying to avoid holding references to internals for a reason, so the burden of proof should be a good reason not to do it. I don't think that applies here, so I've made the change and it was easy.